xapian-letor/api/featurelist_internal.cc

```diff
@@ -0,0 +1,186 @@
+ * USA
+ */
+
+#include <config.h>
+
+#include "xapian-letor/featurelist.h"
+#include "featurelist_internal.h"
+
+#include <cmath>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include "debuglog.h"
+
+using namespace std;
+using namespace Xapian;
+
+void
+FeatureList::Internal::set_database(const Xapian::Database & db)
+{
+    LOGCALL_VOID(API, "FeatureList::set_database", db);
+    featurelist_db = db;
+}
+
+void
+FeatureList::Internal::set_query(const Xapian::Query & query)
+{
+    LOGCALL_VOID(API, "FeatureList::set_query", query);
+    if (query.empty()) {
+	throw Xapian::InvalidArgumentError("Can't initialise with an empty query string");
+    }
+    featurelist_query = query;
+}
+
+void
+FeatureList::Internal::set_doc(const Xapian::Document & doc)
+{
+    LOGCALL_VOID(API, "FeatureList::set_doc", doc);
+    featurelist_doc = doc;
+}
+
+void
+FeatureList::Internal::compute_statistics(const Xapian::Query & letor_query, const Xapian::Database & letor_db,
+					  const Xapian::Document & letor_doc)
+{
+    set_query(letor_query);
+    set_doc(letor_doc);
+    set_database(letor_db);
+
+    compute_termfreq();
+    compute_inverse_doc_freq();
+    compute_doc_length();
+    compute_collection_length();
+    compute_collection_termfreq();
+}
+
+void
+FeatureList::Internal::compute_termfreq()
+{
+    std::map<std::string, long int> tf;
+
+    Xapian::TermIterator docterms = featurelist_doc.termlist_begin();
+    for (Xapian::TermIterator qt = featurelist_query.get_terms_begin(); qt != featurelist_query.get_terms_end(); ++qt) {
+	docterms.skip_to(*qt);
+	if (docterms != featurelist_doc.termlist_end() && *qt == *docterms) {
+	    tf[*qt] = docterms.get_wdf();
+	} else {
+	    tf[*qt] = 0;
+	}
+    }
+    termfreq = tf;
+}
+
+void
+FeatureList::Internal::compute_inverse_doc_freq()
+{
+    std::map<std::string, double> idf;
+
+    for (Xapian::TermIterator qt = featurelist_query.get_terms_begin(); qt != featurelist_query.get_terms_end(); ++qt) {
+	if (featurelist_db.term_exists(*qt)) {
+	    long int totaldocs = featurelist_db.get_doccount();
+	    long int df = featurelist_db.get_termfreq(*qt);
+	    idf[*qt] = log10(totaldocs / (1 + df));
+	} else {
+	    idf[*qt] = 0;
+	}
+    }
+    inverse_doc_freq = idf;
+}
+
+void
+FeatureList::Internal::compute_doc_length()
+{
+    std::map<std::string, long int> len;
+
+    long int temp_count = 0;
+    Xapian::TermIterator dt = featurelist_doc.termlist_begin();
+    // reach the iterator to the start of the title terms i.e. prefix "S"
+    dt.skip_to("S");
+    for ( ; dt != featurelist_doc.termlist_end(); ++dt) {
+	if ((*dt)[0] != 'S') {
+	    // We've reached the end of the S-prefixed terms.
+	    break;
+	}
+	temp_count += dt.get_wdf();
+    }
+    len["title"] = temp_count;
+    len["whole"] = featurelist_db.get_doclength(featurelist_doc.get_docid());
+    len["body"] = len["whole"] - len["title"];
+    doc_length = len;
```

Comment:
Same comments about using the correct types and `std::swap()` to avoid a copy of the map.  Also below...